### PR TITLE
Remove unnecessary dependencies

### PR DESCRIPTION
**Pull request by adwydman**
I removed several dependencies from `package.json`:

 * Set - supported natively in 4.4.7, unused
 * `require-dir` - unused
 * `sealious-datastore-tingo` - usused
 * `sha1` - unused
 * `require-self` - unused

Also `merge` is not needed anymore, because`Object.assign` does pretty much the same thing (supported in 4.4.7). Four files use `merge`: 

 * `config-manager.js` - I rewrote `merge` to `Object.assign` here, because we have tests
 * `resource-type.js` 
 * `resource-type-collection-subject.js`
 * `resource-type-field-subject.js` - all of the three are not changed, because we don't have tests there.

### CODE REVIEW

File: tests/unit-tests/config/config-manager.test.js [hunk 1]
```diff
@@ -24,8 +24,14 @@ describe("ConfigManager", function(){
 	});
 
 	it("gets configuration", function(){
-		const config = ConfigManager.get_configuration("this.key");
-		assert.strictEqual(config instanceof Object, true);
```
Comment (kuba-orlik) [inline tests/unit-tests/config/config-manager.test.js]: Why did you remove these two lines?

**Comment by kuba-orlik** [comment]
1.  `require-self` is necessary for `require("sealious")` to work. Please remove all `require("sealious")` instances and replace them with proper `locreq` calls to specific modules. (`require-self` is a module that is being called during the `prepublish` step of the `npm install` process. See [package.json:8](https://github.com/sealcode/sealious/blob/references/package.json#L8) and the [build status](https://travis-ci.org/sealcode/sealious/builds/149410322))
2.  `sealious-datastore-tingo` *is* used. It is the default datastore. It might not be explicitly `require`d, but Sealious does load it during init time, thanks to the `plugin-manager` magic. I brought it back to the package.json file.
3.  as for the `merge` module: I appreciate the elegance of `Object.assign`, I didn't know such functionality is supported natively by Node.js. But `merge.recursive` and `Object.assign`'s behaviors are different, as demonstrated by this example: 

  ```javascript
  > var objA = {foo: {bar: "baz", bat: "bax"}};
  > var objA = {foo: {bar: "baz", bat: "bax"}};
  
  > merge.recursive(true, objA, objB)
  { foo: { bar: 'baz', bat: 'bax', bam: 'bak' } }
  > Object.assign({}, objA, objB)
  { foo: { bam: 'bak' } }
  ```

  Please revert the Object.assign changes.

PS. remember to `git pull --rebase remove_unnecessary_dependencies`, as I pushed two commits to the branch. (Note the `--rebase`! :) )

File: tests/unit-tests/config/config-manager.test.js (hunk 1 above)
Comment (adwydman) [inline tests/unit-tests/config/config-manager.test.js]: Because they brought no value to the test, I adjusted it to actually get the correct configuration.

**Comment by adwydman** [comment]
Ad 1, Ad 2: Noted.

Ad 3:
 Uhm, I think your example is messed up, 
```
> var objA = {foo: {bar: "baz", bat: "bax"}};
> var objA = {foo: {bar: "baz", bat: "bax"}};
```

These objects are the same.

**Comment by kuba-orlik** [comment]
Yep, sorry, I messed copy&paste :D I've updated my previous comment :)

**Comment by adwydman** [comment]
Ok, I see it now. `Object.assign` overwrites properties to the latest occurance.

**Comment by adwydman** [comment]
Btw, @kuba-orlik did you assign yourself to this PR purposefully?

**Comment by kuba-orlik** [comment]
Yes, as I'm the one who reviews it and will merge it :)

**Comment by adwydman** [comment]
Cool :)

**Comment by kuba-orlik** [comment]
What I find curious is that the unit tests didn't detect a change in behavior from `merge.recursive` to `Object.assign`. Could you add a test or two so they do take that into consideration? :)

**Comment by adwydman** [comment]
Because I used only one level of object nesting, so there was no difference
between recursive merging and `Object.assign`.

2016-08-03 19:09 GMT+02:00 Kuba Orlik <notifications@github.com>:

**Comment by adwydman** [comment]
@kuba-orlik: I've added tests that won't pass if `Object.assign` is used: 

https://github.com/sealcode/sealious/commit/7f82bba2a510928ac5b20030c985f6c724dfd983
https://github.com/sealcode/sealious/commit/f3c95dcb9ed4625281d853f7178ef22040ef3499

**Comment by kuba-orlik** [comment]
With `require-self` removed, `require("sealious")` *will not work*. In the `remove_unnecessary_dependencies` branch there are still many instances of `require("sealious")`. I will merge it once they are all gone :)

Here's a list of files that demand that still `require("sealious")`

```
lib/utils/run-action.js
lib/plugins/plugin-manager.js
lib/core-services/file-manager.js
lib/base-chips/field-types/file.js
lib/base-chips/field-types/single_reference.js
lib/base-chips/field-types/username.js
lib/base-chips/access-strategy-types/super.js
lib/base-chips/access-strategy-types/themselves.js
lib/base-chips/access-strategy-types/or.js
lib/base-chips/access-strategy-types/not.js
lib/base-chips/access-strategy-types/and.js
lib/base-chips/resource_type.user.js
lib/action.js
tests/unit-tests/response/response.test.js
lib/context.js
tests/unit-tests/base-chips/field-types/file.test.js
```

**Comment by adwydman** [comment]
There's a separate branch for that: https://github.com/sealcode/sealious/tree/remove_require_sealious :)

**Comment by kuba-orlik** [comment]
I know :) I hope you see that I can't merge this one unless the other one is finished.

**Comment by adwydman** [comment]
There are no `require('sealious')` there. Do you want me to merge that branch with this one?

**Comment by kuba-orlik** [comment]
Correct my method if it's wrong, but this is the way I check whether or not there are `require("sealious")` instances in this branch: 

```
kuba@laptop:~/projects/sealcode/sealious/sealious$ git checkout remove_unnecessary_dependencies 
Switched to branch 'remove_unnecessary_dependencies'
Your branch is up-to-date with 'origin/remove_unnecessary_dependencies'.
kuba@laptop:~/projects/sealcode/sealious/sealious$ git pull
Already up-to-date.
kuba@laptop:~/projects/sealcode/sealious/sealious$ ag 'require\("sealious"\)' -c
lib/utils/run-action.js:1
lib/plugins/plugin-manager.js:1
lib/core-services/file-manager.js:1
lib/context.js:2
lib/base-chips/field-types/file.js:1
lib/base-chips/field-types/single_reference.js:1
lib/base-chips/field-types/username.js:1
lib/base-chips/access-strategy-types/super.js:1
lib/base-chips/access-strategy-types/themselves.js:1
lib/base-chips/access-strategy-types/or.js:1
lib/base-chips/access-strategy-types/not.js:1
lib/base-chips/access-strategy-types/and.js:1
lib/base-chips/resource_type.user.js:1
lib/action.js:1
tests/unit-tests/response/response.test.js:1
tests/unit-tests/base-chips/field-types/file.test.js:1
kuba@laptop:~/projects/sealcode/sealious/sealious$ ag 'require\("sealious"\)' 
lib/utils/run-action.js
2:const Sealious = require("sealious");

lib/plugins/plugin-manager.js
2:const Sealious = require("sealious");

lib/core-services/file-manager.js
7:const Sealious = require("sealious");

lib/context.js
52:		const Sealious = require("sealious");
73:	const Sealious = require("sealious");

lib/base-chips/field-types/file.js
2:const Sealious = require("sealious");

lib/base-chips/field-types/single_reference.js
2:const Sealious = require("sealious");

lib/base-chips/field-types/username.js
3:const Sealious = require("sealious");

lib/base-chips/access-strategy-types/themselves.js
2:const Sealious = require("sealious");

lib/base-chips/access-strategy-types/or.js
2:const Sealious = require("sealious");

lib/base-chips/access-strategy-types/super.js
2:const Sealious = require("sealious");

lib/base-chips/access-strategy-types/not.js
2:const Sealious = require("sealious");

lib/base-chips/access-strategy-types/and.js
2:const Sealious = require("sealious");

lib/base-chips/resource_type.user.js
2:const Sealious = require("sealious");

lib/action.js
2:const Sealious = require("sealious");

tests/unit-tests/response/response.test.js
4:const Sealious = require("sealious");

tests/unit-tests/base-chips/field-types/file.test.js
4:const Sealious = require("sealious");
```

**Comment by kuba-orlik** [comment]
Also, this is the state of one of the files on the tip of this branch: https://github.com/sealcode/sealious/blob/369e5384a0a3cc1663c37cab6a1971c59406843b/lib/base-chips/access-strategy-types/or.js#L2

**Comment by kuba-orlik** [comment]
(Maybe you meant to merge the `remove_require_sealious` branch to this one? :) )

**Comment by adwydman** [comment]
You are wrong, because you are checking `remove_unnecessary_dependencies` and not `remove_require_sealious` ;)

**Comment by kuba-orlik** [comment]
But... that's what this PR is for, right? The `remove_unnecessary_dependencies` branch :)

selection_112

So I thought the request was: "Please merge `remove_unnecessary_dependencies` branch into `references`". And I cannot, because it's not ready. If you merge `remove_require_sealious` into `remove_unnecessary_dependencies`, or create a separate PR for `remove_require_sealious`, that's different business. 

Excuse me if I missed something in what you were communicating, but I was sure you were talking about the branch that this PR is about.

**Comment by adwydman** [comment]
Ok, miscommunication. I'll merge `remove_require_sealious` to this branch.

**Comment by adwydman** [comment]
Merged `remove_require_sealious` into this one.

**Comment by kuba-orlik** [comment]
Two more files left :) 

```
tests/unit-tests/response/response.test.js:1
tests/unit-tests/base-chips/field-types/file.test.js:1
```

I don't think removing `require("sealious")` from them would break anything, hm?

**Comment by adwydman** [comment]
You won't know unless you try it. :)

**Comment by kuba-orlik** [comment]
Integration tests [fail](https://travis-ci.org/sealcode/sealious/builds/151517929#L566) on this branch, but [pass](https://travis-ci.org/sealcode/sealious/builds/151516366) on `references`. 

They fail here because [File Manager receives only the datastore *type*, not the actual datastore instance](https://github.com/sealcode/sealious/pull/287#issuecomment-237893817) -  [proof](https://travis-ci.org/sealcode/sealious/builds/151517929#L566). @adwydman Will you look into that? :)

**Comment by adwydman** [comment]
@kuba-orlik: The tests pass on Travis, but it also executes `node .` which fails.

**Comment by kuba-orlik** [comment]
`node . ` runs integration tests ;)

**Comment by adwydman** [comment]
Okay! But I don't know how to instantiate Datastore without using `require('sealious')`.

**Comment by kuba-orlik** [comment]
Hmm we could make Datastore a part of `context`, what do you think about that?

**Comment by kuba-orlik** [comment]
https://github.com/sealcode/sealious/blob/no_accept_reject_interface/lib/subject/subject-types/resource-type-collection-subject.js#L44 :)
